File: misc/leetcode/freedom-trail.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Solution:
    def findRotateSteps(self, ring, key):
        """
        :type ring: str
        :type key: str
        :rtype: int
        """

        n = len(ring)
        m = len(key)
        dp = [[0] * n, [0] * n]
        now = 0
        for i in range(n):
            dp[now][i] = i

        for i in range(m):
            for j in range(n):
                res = 1 << 30
                match = False

                for step in range(0, n):
                    k = (j + step) % n
                    if ring[k] == key[i]:
                        match = True
                    if match:
                        res = min(res, dp[now][k] + step)

                match = False
                for step in range(0, n):
                    k = (j + n - step) % n
                    if ring[k] == key[i]:
                        match = True
                    if match:
                        res = min(res, dp[now][k] + step)

                dp[1 - now][j] = res
            now = 1 - now

        ans = min(dp[now]) + m
        return ans


if __name__ == '__main__':
    sol = Solution()
    print(sol.findRotateSteps('godding', 'gd'))
    print(sol.findRotateSteps(
        "sqlfcudiojjzmdmvbqgtkudggbazwtqgzrbxlooxcfnvzkvyvrroakdhnwcfyzyefiuatefegiragiqdrggictalanfupkuvjyid",
        "jogduakfgovnolkbjwelatfgfunqgvajvwtrzguyydiqaucqrzzcuhxcpkilfebqyytaxikigemzatzgmcdbodriddnrvgffsriv"))

File: leetcode/cpp/freedom-trail.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <algorithm>
#include <map>
#include <set>
#include <string>
#include <vector>
using namespace std;

class Solution {
   public:
    int findRotateSteps(string ring, string key) {
        int n = ring.size();
        int m = key.size();
        vector<int> dp[2];
        dp[0].resize(n);
        dp[1].resize(n);
        int now = 0;
        for (int i = 0; i < n; i++) {
            dp[now][i] = i;
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int res = INT_MAX;
                bool ok = false;

                for (int step = 0; step < n; step++) {
                    int k = (step + j) % n;
                    if (ring[k] == key[i]) ok = true;
                    if (ok) {
                        res = min(res, dp[now][k] + step);
                    }
                }

                ok = false;
                for (int step = 0; step < n; step++) {
                    int k = (j + n - step) % n;
                    if (ring[k] == key[i]) ok = true;
                    if (ok) {
                        res = min(res, dp[now][k] + step);
                    }
                }

                dp[1 - now][j] = res;
            }
            now = 1 - now;
        }
        int ans = INT_MAX;
        for (int i = 0; i < n; i++) {
            ans = min(ans, dp[now][i]);
        }
        ans += m;
        return ans;
    }
};